design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1072 @@
+# KLIP 3 - Deserialization and Derialization of Single Fields.
+
+**Author**: @big-andy-coates | 
+**Release Target**: TBD | 
+**Status**: In Discussion | 
+**Discussion**: 
+
+**tl;dr:** _Add the ability to deserialize from, and serialize to, primitive types, 
+arrays and maps._
+
+## Terminology
+
+Both JSON and AVRO support serialization of objects with multiple named fields.
+JSON has _JSON objects_, while Avro has _Avro records_.
+
+Both formats also support other types: JSON supports an array type and a small 
+selection of primitive types (`Number`, `String`, and `Boolean`); Avro supports
+array and map types along with a wider range of primitive types.
+
+The collective term _record_ is used to refer to JSON _objects_ and Avro _records_. 
+The term _non-record_ is used to refer to any other types.
+
+A row in KSQL has a logical schema made up of multiple fields. Fields within the row's
+schema are persisted to either the key or value of the underlying Kafka message. 
+In this sense the key and value also have a logical schema.  
+
+The term _single field_ is used to describe a key or value logical schema where 
+the schema contains only a single field. 
+           
+## Motivation and background
+
+Currently, the `JSON` and `AVRO` formats in KSQL can only deserialize _record_ values, 
+i.e. those in the form of a JSON object or an Avro record. These formats can 
+only deserialize _non-record_ types if the type is contained as a field within a
+top-level _record_ type.
+
+Likewise, KSQL always serializes the value fields with in a schema as fields within
+a top-level _record_. There is no support for serializing schemas that contain only 
+a single value field as _non-record_ types. For example, if a query selects only a 
+single `INT` value field, it will be serialized within a _record_, it can not be 
+serialized as a simple integer.
+
+On the other hand, KSQL only supports string keys, i.e. a _non-record_ type.
+
+Being able to both process and produce both _record_ and _non-record_ values opens up 
+KSQL to more use cases, and is a precondition of enhancing KSQL to support _record_
+keys, a.k.a. structured keys. (Support for _non_record_ keys will be needed to 
```

Comment:
suggest "aka structured keys" -> "akak _structured_ or _composite_ keys"

Reply:
Done, though I wasn't aware of anyone calling them `composite` keys.